lib/utils/ajvErrorMessageBuilder.js

```diff
@@ -0,0 +1,42 @@
+'use strict';
+
+/*
+ * For error object structure, see https://github.com/epoberezkin/ajv#error-objects
+ */
+const buildErrorMessages = errors => {
+  errors = addUserFriendlyMessage(errors);
+  errors = errors.sort((a, b) => b.dataPath.length - a.dataPath.length);
+  if (!errors.length) throw new Error('Validation errors array cannot be emptry');
+  return errors[0].friendlyMessage
+    ? [errors[0].friendlyMessage]
+    : [`Error: ${errors[0].dataPath} ${errors[0].message}`];
+};
+
+const addUserFriendlyMessage = errors => {
+  return errors.map(error => {
+    let friendlyMessage;
+
+    // regex matches '.functions['xxx'].events[x].xxx'
+    const isEventParam = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]\.[\w]+$/.test(error.dataPath);
+    if (error.params && error.params.additionalProperty && isEventParam) {
+      const eventName = error.dataPath.split('.')[3];
+      friendlyMessage = `Unsupported parameter '${error.params.additionalProperty}' for '${eventName}' event`;
+    }
+
+    // regex matches '.functions['xxx'].events[x]'
+    const isFunctionEvent = /^\.functions\['[\w-_]+'\]\.events\[[0-9]+\]$/.test(error.dataPath);
+    if (error.params && error.params && error.params.additionalProperty && isFunctionEvent) {
+      friendlyMessage = `Unsupported function event '${error.params.additionalProperty}'`;
+    }
+    if (friendlyMessage) {
+      error.friendlyMessage = friendlyMessage;
```

Comment:
It'll be nice to also ensure that all other errors reported by AJV, which strictly relate to fact that unsupported event name was used, are not output (having them out, will be really noisy)

If there are no easy means to detect and filter those errors, I would then just throw this only error in such case.

Reply:
Let me share how I ended up with the friendly message implementation and the general handling of error messages. 

It always shows only one error message. 

With one error message, the task becomes how to find the most relevant one. The tricky thing is that friendly message error may not always be the most relevant one. Some times AJV produces error objects, where even two friendly message types can fit and still those friendly messages would be irrelevant. 

Here's example. Let's say some function's event property is expected to be an integer. If the user accidentally provides a string, then AJV creates error objects that would technically apply to a friendly message pattern. However, the expected message would be the most specific one.

I've included the above example in the test. Check this test out:
https://github.com/serverless/serverless/blob/23ff0510a727c84410141b3c185736cacd560dee/lib/utils/ajvErrorMessageBuilder.test.js#L100

What I noticed is that the most specific (and the most relevant) error appears to have the longest `error.dataPath`. So out of all error objects we choose the object with the longest `dataPath` param. If such an error object appears to fit a friendly message pattern, then friendly message is shown. Otherwise, the default error message is crafted out of error object params.

Such approach may sound not ideal and may need fine-tuning (for example in case of equal data paths, which error should be the most relevant?). However, as I use it, this seem to work pretty good.

Let me know what you think about it

Reply:
@preshetin that sounds good.

Ideally I think we should:
1. Show only one error message for single issue (and best if most DX friendly one is chosen or prepared)
2. Report all issues in config.

If it appears difficult to achieve above (I would definitely not make implementation too complex, as then it'll likely become a maintenance burden, and we'd open a pandora's box of issues to address in future), then I would value (1) over (2). So if parsing of other messages appears difficult, just show error for single issue.

Reply:
Yes, currently the first (1) approach is implemented. It shows only one error message and tries to show the most relevant error message. 

Of course, some fine-tuning may be required as we face cases when the error message is unclear. Such cases, I think, can be dealt with in follow up PRs.